src/solver/solver.cxx

```diff
@@ -491,14 +491,66 @@ int Solver::solve(int NOUT, BoutReal TIMESTEP) {
                                    .doc("Write initial state as time point 0?")
                                    .withDefault(!restart || !append);
 
-  if (dump_on_restart) {
+  // Run RHS once to ensure all variables set
+  if (run_rhs(simtime)) {
+    throw BoutException("Physics RHS call failed\n");
+  }
 
-    /// Write initial state as time-point 0
+  const bool error_on_unused_options =
+      globaloptions["input"]["error_on_unused_options"]
+          .doc(
+              "Error if there are any unused options before starting the main simulation")
+          .withDefault(true);
+
+  const bool validate_input =
+      globaloptions["input"]["validate"]
+          .doc(
+              "Check for unused options and stop")
+          .withDefault(false);
+
+  if (error_on_unused_options) {
+    Options unused = globaloptions.getUnused();
+    if (not unused.getChildren().empty()) {
```

Comment:
Maybe this whole chunk should be broken out into a function somewhere else? I'm still not keen on this living in `Solver::solve`, but moving it out requires something like `BOUTMAIN` calling `PhysicsModel::init/run_rhs` to get all the options.

Also the `input:help` functionality needs to essentially be called in the same place, and then we may have to be careful that nothing has tried to write to any files, or we'll end up clobbering stuff. I was wondering if we could somehow run in a "fake" location somewhere to avoid this. We'd still need access to the input file and grid file, etc.

Reply:
It might be nice to have the init and first RHS call be done in a separate call, rather than inside `Solver`. I remember a long time ago (2013?) we had to change how Solver initialised and ran, because of the weird interface PETSc had then. Perhaps its worth revisiting?